pkg/ext/store.go

```diff
@@ -2,92 +2,260 @@ package ext
-// up to the implementer. For example, resources could be stored in another ETCD
-// database, in a SQLite database, in another built-in resource such as Secrets.
-// It is also possible to have no storage at all.
+// It converts an object or a list of objects to a table, which is used by kubectl
+// (and Rancher UI) to display a table of the items.
 //
-// Errors returned by the Store should use errors from k8s.io/apimachinery/pkg/api/errors. This
-// will ensure that the right error will be returned to the clients (eg: kubectl, client-go) so
-// they can react accordingly. For example, if an object is not found, store should
-// return the following error:
+// Currently, we use the default table convertor which will show two columns: Name and Created At.
+func ConvertToTable[T runtime.Object](ctx context.Context, object runtime.Object, tableOptions runtime.Object, groupResource schema.GroupResource, columnDefs []metav1.TableColumnDefinition, convertFn ConvertFunc[T]) (*metav1.Table, error) {
+	result, err := convertToTable(ctx, object, tableOptions, groupResource, columnDefs, convertFn)
+	if err != nil {
+		return nil, convertError(err)
+	}
+	return result, nil
+}
+
+func ConvertToTableDefault[T runtime.Object](ctx context.Context, object runtime.Object, tableOptions runtime.Object, groupResource schema.GroupResource) (*metav1.Table, error) {
+	return ConvertToTable[T](ctx, object, tableOptions, groupResource, nil, nil)
+}
+
+func convertToTable[T runtime.Object](ctx context.Context, object runtime.Object, tableOptions runtime.Object, groupResource schema.GroupResource, columnDefs []metav1.TableColumnDefinition, convertFn ConvertFunc[T]) (*metav1.Table, error) {
+	defaultTableConverter := rest.NewDefaultTableConvertor(groupResource)
+	table, err := defaultTableConverter.ConvertToTable(ctx, object, tableOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	if columnDefs == nil {
+		return table, nil
+	}
+
+	// Override only if there were definitions before (to respect the NoHeader option)
+	if len(table.ColumnDefinitions) > 0 {
+		table.ColumnDefinitions = columnDefs
+	}
+	table.Rows = []metav1.TableRow{}
+	fn := func(obj runtime.Object) error {
+		objT, ok := obj.(T)
+		if !ok {
+			var zeroT T
+			return fmt.Errorf("expected %T but got %T", zeroT, obj)
+		}
+		cells := convertFn(objT)
+		if len(cells) != len(columnDefs) {
+			return fmt.Errorf("defined %d columns but got %d cells", len(columnDefs), len(cells))
+		}
+
+		table.Rows = append(table.Rows, metav1.TableRow{
+			Cells:  cellStringToCellAny(cells),
+			Object: runtime.RawExtension{Object: obj},
+		})
+		return nil
+	}
+	switch {
+	case meta.IsListType(object):
+		if err := meta.EachListItem(object, fn); err != nil {
+			return nil, err
+		}
+	default:
+		if err := fn(object); err != nil {
+			return nil, err
+		}
+	}
+
+	return table, nil
+}
+
+func cellStringToCellAny(cells []string) []any {
+	res := []any{}
```

Comment:
nit: Maybe do this to avoid returning an empty slice instead of a nil slice when the input slice is nil.
```suggestion
	var res []any
```

Reply:
Sure